Upgrade to PHPUnit v10

lcobucci opened the pull request:
Fix #740 

This makes the test suite compatible with the latest PHPUnit version and upgrades everything to it, which removes the SEGFAULTs we were having previously.

(the first commit is from #737, a rebase to sync is more than enough to remove it from the last merged branch)

lcobucci committed:
Ensure GPG isn't used on tests

Some people - like me - have a tweaked/opinionated global Git
configuration that forces GPG signatures everywhere, which basically
breaks the tests.

This standardises the expected configuration for tests, preventing
unexpected breakages.

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci committed:
Ensure data providers are static

PHPUnit 10.x only accepts static data providers, so...

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci committed:
Move away from deprecated MockObject#withConsecutive()

More info: https://github.com/sebastianbergmann/phpunit/issues/5063

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci committed:
Upgrade to PHPUnit v10

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci commented:
Replaces #736

Ocramius commented:
Argh, this is possibly blocked by https://github.com/laminas/laminas-ci-matrix-action/pull/189 and https://github.com/laminas/laminas-continuous-integration-action/issues/130

lcobucci commented:
> Argh, this is possibly blocked by https://github.com/laminas/laminas-ci-matrix-action/pull/189 and https://github.com/laminas/laminas-continuous-integration-action/issues/130

Fun!

We can "hide" the SEGFAULTs by temporarily relying on `exec()`.

Trade-offs, eh!?

lcobucci commented:
Can't we add it as extension to be installed for this package (at least until that addressed on the upstream)? 

Alternatively we can also use a pre-install script to run arbitrary commands as root: https://github.com/laminas/laminas-continuous-integration-action/blob/1.33.x/entrypoint.sh#L162

That might help to only install a coverage driver for specific jobs and allow moving this forward.

Ocramius commented:
I really just need to do the upstream releasing/fixing: however you put the problem, the blocker is me and my constantly clogged schedule :S

Ocramius commented:
I'd be fine with using `exec()` BTW: the code is well tested anyway, so it's an implementation change only.

lcobucci commented:
> I'd be fine with using `exec()` BTW: the code is well tested anyway, so it's an implementation change only.

I'll send a different PR to tackle this, I've the changes in my local history 👍

lcobucci committed:
Override Infection CI job definition

Given the plans for laminas-continuous-integration-action v2 is to always
ship PCOV, we can simply enable it here to allow us to detach the two
processes.

This temporarily disables the default infection job and provides a new
configuration that uses PCOV instead.

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci commented:
> > I'd be fine with using `exec()` BTW: the code is well tested anyway, so it's an implementation change only.
> 
> I'll send a different PR to tackle this, I've the changes in my local history +1

Scratch that, we can already enable PCOV here and solve the issue.

I've pushed an extra commit and rebased the baseline PR to verify that CI works for both.

@Ocramius @asgrim please approve the workflow run (whenever) on both PRs to make sure things are fine :+1:

lcobucci commented:
Silly me... I've overlooked the generated command for infection on the matrix :facepalm: 
I used `vendor/bin/infection` to test, CI is using `phpdbg -qrr ./vendor/bin/roave-infection-static-analysis-plugin`

Well, that would never work :rofl:

lcobucci commented:
All the things I can think of are horrendous... let's release CI-action v2 :laughing:

Ocramius commented:
Yeah, and `phpdbg` + `pcov`  together do lead to segfaults :S

Ocramius commented:
@lcobucci doing some paid work, then picking up OCI-CatalogService again

lcobucci commented:
@Ocramius should we close this or bump to v11? 😅

Ocramius commented:
If it works, we can merge here, although no idea if it does :P  

My OSS efforts have been almost zero this year, sorry.

lcobucci commented:
@Ocramius I had some time here... shall we try again?

lcobucci commented:
@Ocramius all the checks should be fine, except for Infection as it still requires laminas-ci-matrix-action 2 to be released (drop phpdbg and use pcov).

Ocramius commented:
@lcobucci should we perhaps skip running it via `laminas-ci-matrix-*` there, and run `roave/infection-static-analysis-plugin` (locked) ourselves?

lcobucci commented:
@Ocramius I'm not that familiar with the config there, can you check if https://github.com/Roave/BackwardCompatibilityCheck/pull/741/commits/7d3abac7d279e82ffc72bcf2c4092c6e0891cc21 handles it?

Ocramius commented:
@lcobucci see https://github.com/laminas/laminas-ci-matrix-action/blob/7f7976e6faf1ee8fcfca77cf0487cc4d0b486f11/laminas-ci.schema.json

Ocramius commented:
I managed to get things running in a similar way here: https://github.com/Ocramius/DoctrineBatchUtils/pull/392

That schema file is a blessing :D

lcobucci commented:
@Ocramius the schema file helps a lot :+1:

I took a slightly different approach so that we always run against the minimum supported PHP version (according to https://github.com/laminas/laminas-ci-matrix-action?tab=readme-ov-file#job-element it should work).

I **think** everything will just work :smile:

Ocramius commented:
> I think everything will just work 😄

We do this not because it is easy, but because we believed it would be easy.

Ocramius commented:
:ship: thanks @lcobucci!
